drivers/spi/spi_silabs_eusart.c

```diff
@@ -199,35 +592,145 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
 	struct spi_context *ctx = &data->ctx;
+	int ret;
+
+	size_t total_padded_transaction_size = spi_silabs_longest_transfer_size(ctx);
+
+	spi_silabs_eusart_clear_txrx_fifos(eusart_config->base);
+
+	ret = spi_silabs_prepare_dma_transaction(dev, total_padded_transaction_size);
+
+	if (ret) {
+		return ret;
+	}
+
+	spi_silabs_eusart_pm_policy_get(dev);
+
+	spi_context_cs_control(ctx, true);
+
+	if (total_padded_transaction_size == 0U) {
+		goto force_transaction_close;
+	}
+
+	/* RX channel needs to be ready before TX channel actually starts */
+	ret = dma_start(data->dma_channel_rx.dma_dev, data->dma_channel_rx.channel_nb);
+
+	if (ret) {
+		goto force_transaction_close;
+	}
+
+	ret = dma_start(data->dma_channel_tx.dma_dev, data->dma_channel_tx.channel_nb);
+
+	if (ret) {
+		goto force_transaction_close;
+	}
+
+	if (spi_context_wait_for_completion(&data->ctx) >= 0) {
+		/* Successful transaction */
+		return 0;
+	}
+
+force_transaction_close:
+	spi_context_cs_control(ctx, false);
+	spi_silabs_eusart_pm_policy_put(dev);
+	return ret;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+
+static int spi_silabs_eusart_xfer_polling(const struct device *dev, const struct spi_config *config)
+{
 	const struct spi_silabs_eusart_config *eusart_config = dev->config;
+	struct spi_silabs_eusart_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret = 0;
 
+	spi_silabs_eusart_pm_policy_get(dev);
 	spi_context_cs_control(ctx, true);
 
-	do {
-		ret = spi_silabs_eusart_shift_frames(eusart_config->base, data);
-	} while (!ret && spi_silabs_eusart_transfer_ongoing(data));
+	if (spi_silabs_eusart_transfer_ongoing(data)) {
+		do {
+			ret = spi_silabs_eusart_shift_frames(eusart_config->base, data);
+		} while (!ret && spi_silabs_eusart_transfer_ongoing(data));
+	}
 
 	spi_context_cs_control(ctx, false);
 	spi_context_complete(ctx, dev, 0);
+
+	spi_silabs_eusart_pm_policy_put(dev);
+	return ret;
+}
+
+static int spi_silabs_eusart_transceive(const struct device *dev,
+					const struct spi_config *config,
+					const struct spi_buf_set *tx_bufs,
+					const struct spi_buf_set *rx_bufs,
+					bool asynchronous,
+					spi_callback_t cb,
+					void *userdata)
+{
+	struct spi_silabs_eusart_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret;
+
+	spi_context_lock(ctx, asynchronous, cb, userdata, config);
+
+	ret = spi_silabs_eusart_configure(dev, config);
+
+	if (ret) {
+		goto out;
+	}
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, 1);
+
+	if (is_dma_enabled_instance(dev)) {
+		/* DMA transfer handle a/synchronous transfers */
+		ret = spi_silabs_eusart_xfer_dma(dev, config);
+	} else {
+		if (asynchronous) {
+			/* Asynchronous transfers without DMA is not implemented,
+			 * please configure the device tree
+			 * instance with the proper DMA configuration.
+			 */
+			ret = -ENOTSUP;
+			goto out;
+		} else {
+			ret = spi_silabs_eusart_xfer_polling(dev, config);
+		}
+	}
+
+	ret = spi_context_wait_for_completion(ctx);
```

Comment:
The DMA case already calls this within `spi_silabs_eusart_xfer_dma`, doing it again here doesn't seem right.